File: repositories/open-agentic-platform/product/packages/skill-command-factory/src/parser.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Feature: 071-skill-command-factory

/**
 * Frontmatter parser for skill markdown files.
 *
 * FR-001: Parse YAML frontmatter with skill-specific fields.
 * FR-009: Invalid frontmatter produces a warning, does not block other skills.
 * Contract note: Files without frontmatter are prompt-type with allowed_tools: "*".
 */

import { parse as parseYaml } from "yaml";
import type {
  ParsedSkill,
  SkillFrontmatter,
  SkillHookDeclaration,
  SkillLoadResult,
  SkillType,
} from "./types.js";
import { basename } from "node:path";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/;
const VALID_SKILL_TYPES: readonly SkillType[] = ["prompt", "agent", "headless"];
const VALID_HOOK_HANDLER_TYPES = new Set(["bash", "agent", "prompt"]);

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Parse a single skill markdown file into a ParsedSkill.
 *
 * Returns a SkillLoadResult with status "ok", "warning", or "error".
 */
export function parseSkillFile(
  content: string,
  filePath: string,
): SkillLoadResult {
  const match = content.match(FRONTMATTER_RE);

  if (!match) {
    // Backward compatibility: no frontmatter → prompt skill, all tools, name from filename
    return backwardCompatSkill(content, filePath);
  }

  const [, yamlBlock, body] = match;

  let raw: Record<string, unknown>;
  try {
    raw = parseYaml(yamlBlock) as Record<string, unknown>;
    if (raw === null || typeof raw !== "object") {
      return backwardCompatSkill(content, filePath);
    }
  } catch (err) {
    return {
      filePath,
      status: "warning",
      message: `Invalid YAML frontmatter: ${err instanceof Error ? err.message : String(err)}`,
    };
  }

  // Validate and extract frontmatter fields
  const validation = validateFrontmatter(raw, filePath);
  if (validation.error) {
    return {
      filePath,
      status: "warning",
      message: validation.error,
    };
  }

  const fm = validation.frontmatter!;

  const skill: ParsedSkill = {
    name: fm.name,
    description: fm.description,
    skillType: fm.type,
    allowedTools: fm.allowed_tools,
    model: fm.model,
    hooks: fm.hooks ?? {},
    trigger: fm.trigger ?? null,
    body: body.trim(),
    sourcePath: filePath,
  };

  return { filePath, status: "ok", skill };
}

/**
 * Derive a skill name from a file path.
 * E.g., "/path/to/commit.md" → "commit"
 */
export function skillNameFromPath(filePath: string): string {
  return basename(filePath, ".md");
}

// ---------------------------------------------------------------------------
// Backward compatibility (Contract note)
// ---------------------------------------------------------------------------

function backwardCompatSkill(
  content: string,
  filePath: string,
): SkillLoadResult {
  const name = skillNameFromPath(filePath);
  const skill: ParsedSkill = {
    name,
    description: `Skill loaded from ${basename(filePath)}`,
    skillType: "prompt",
    allowedTools: "*",
    hooks: {},
    trigger: null,
    body: content.trim(),
    sourcePath: filePath,
  };
  return { filePath, status: "ok", skill };
}

// ---------------------------------------------------------------------------
// Frontmatter validation
// ---------------------------------------------------------------------------

interface ValidationResult {
  frontmatter?: SkillFrontmatter;
  error?: string;
}

function validateFrontmatter(
  raw: Record<string, unknown>,
  filePath: string,
): ValidationResult {
  // Name: required (R-002 says minimal required field)
  const name =
    typeof raw.name === "string" && raw.name.trim() !== ""
      ? raw.name.trim()
      : skillNameFromPath(filePath);

  // Description: optional, defaults to generic
  const description =
    typeof raw.description === "string" && raw.description.trim() !== ""
      ? raw.description.trim()
      : `Skill loaded from ${basename(filePath)}`;

  // Type: optional, defaults to "prompt"
  let type: SkillType = "prompt";
  if (typeof raw.type === "string") {
    if (!VALID_SKILL_TYPES.includes(raw.type as SkillType)) {
      return {
        error: `Invalid skill type "${raw.type}"; expected one of: ${VALID_SKILL_TYPES.join(", ")}`,
      };
    }
    type = raw.type as SkillType;
  }

  // Allowed tools: optional, defaults to "*"
  let allowed_tools: readonly string[] | "*" = "*";
  if (raw.allowed_tools !== undefined) {
    if (raw.allowed_tools === "*") {
      allowed_tools = "*";
    } else if (Array.isArray(raw.allowed_tools)) {
      const tools = raw.allowed_tools.filter(
        (t): t is string => typeof t === "string" && t.trim() !== "",
      );
      if (tools.length === 0) {
        return { error: "allowed_tools array is empty" };
      }
      allowed_tools = tools;
    } else {
      return {
        error: `allowed_tools must be "*" or an array of tool names`,
      };
    }
  }

  // Model: optional string
  const model =
    typeof raw.model === "string" && raw.model.trim() !== ""
      ? raw.model.trim()
      : undefined;

  // Hooks: optional map of event → handler declarations (FR-008)
  let hooks: Record<string, SkillHookDeclaration[]> | undefined;
  if (raw.hooks !== undefined && raw.hooks !== null) {
    if (typeof raw.hooks !== "object" || Array.isArray(raw.hooks)) {
      return { error: "hooks must be an object mapping event types to handler arrays" };
    }
    hooks = {};
    for (const [event, handlers] of Object.entries(
      raw.hooks as Record<string, unknown>,
    )) {
      if (!Array.isArray(handlers)) {
        return { error: `hooks.${event} must be an array of handler declarations` };
      }
      hooks[event] = [];
      for (const h of handlers) {
        if (typeof h !== "object" || h === null) {
          return { error: `hooks.${event} contains an invalid handler` };
        }
        const hObj = h as Record<string, unknown>;
        if (typeof hObj.name !== "string" || hObj.name.trim() === "") {
          return { error: `hooks.${event} handler missing required "name" field` };
        }
        if (
          typeof hObj.type !== "string" ||
          !VALID_HOOK_HANDLER_TYPES.has(hObj.type)
        ) {
          return {
            error: `hooks.${event}.${hObj.name} has invalid type; expected: bash, agent, prompt`,
          };
        }
        if (typeof hObj.run !== "string" || hObj.run.trim() === "") {
          return {
            error: `hooks.${event}.${hObj.name} missing required "run" field`,
          };
        }
        hooks[event].push({
          name: hObj.name as string,
          type: hObj.type as "bash" | "agent" | "prompt",
          if: typeof hObj.if === "string" ? hObj.if : undefined,
          run: hObj.run as string,
        });
      }
    }
  }

  // Trigger: optional string or null
  const trigger =
    typeof raw.trigger === "string" && raw.trigger.trim() !== ""
      ? raw.trigger.trim()
      : null;

  return {
    frontmatter: {
      name,
      description,
      type,
      allowed_tools,
      model,
      hooks,
      trigger,
    },
  };
}

File: repositories/open-agentic-platform/crates/skill-factory/src/parser.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Bartek Kus

//! YAML frontmatter extraction and validation (FR-001, FR-009).
//!
//! Delegates YAML parsing to `agent-frontmatter` (spec 054) and converts
//! `UnifiedFrontmatter` to `SkillFrontmatter` for backward compatibility.

use crate::types::{AllowedTools, ParsedSkill, SkillFrontmatter, SkillLoadResult, SkillType};
use regex::Regex;
use std::path::Path;
use std::sync::LazyLock;

/// Regex for `---\n<yaml>\n---\n<body>` frontmatter blocks.
static FRONTMATTER_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?s)^---\r?\n(.*?)\r?\n---\r?\n?(.*)$").unwrap());

/// Parse a skill `.md` file from its contents and path.
///
/// - Files with valid YAML frontmatter are fully parsed via `agent-frontmatter` (FR-001).
/// - Files without frontmatter are backward-compatible: treated as prompt-type
///   with `allowed_tools: *` and name derived from filename (Contract note).
/// - Invalid frontmatter produces an `Error` result but does not prevent other
///   skills from loading (FR-009).
pub fn parse_skill_file(content: &str, path: &Path) -> SkillLoadResult {
    match FRONTMATTER_RE.captures(content) {
        Some(caps) => {
            let yaml_str = caps.get(1).unwrap().as_str();
            let body = caps.get(2).unwrap().as_str().to_string();

            // Delegate YAML parsing to agent-frontmatter, then convert.
            match serde_yaml::from_str::<SkillFrontmatter>(yaml_str) {
                Ok(fm) => match validate_frontmatter(&fm) {
                    Ok(()) => SkillLoadResult::Ok(ParsedSkill {
                        frontmatter: fm,
                        body,
                        source_path: path.to_path_buf(),
                    }),
                    Err(warning) => SkillLoadResult::Warning {
                        skill: ParsedSkill {
                            frontmatter: fm,
                            body,
                            source_path: path.to_path_buf(),
                        },
                        message: warning,
                    },
                },
                Err(e) => SkillLoadResult::Error {
                    path: path.to_path_buf(),
                    message: format!("invalid YAML frontmatter: {e}"),
                },
            }
        }
        None => {
            // Backward compatibility: no frontmatter → prompt skill with all tools.
            let name = skill_name_from_path(path);
            SkillLoadResult::Ok(ParsedSkill {
                frontmatter: SkillFrontmatter {
                    name,
                    description: None,
                    skill_type: SkillType::Prompt,
                    allowed_tools: AllowedTools::all(),
                    model: None,
                    hooks: Default::default(),
                    trigger: None,
                },
                body: content.to_string(),
                source_path: path.to_path_buf(),
            })
        }
    }
}

/// Derive a skill name from the file path (strip directory and `.md` extension).
pub fn skill_name_from_path(path: &Path) -> String {
    path.file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("unnamed")
        .to_string()
}

/// Validate parsed frontmatter fields. Returns `Ok(())` if clean,
/// or `Err(warning_message)` for non-fatal issues.
fn validate_frontmatter(fm: &SkillFrontmatter) -> Result<(), String> {
    let mut warnings = Vec::new();

    if fm.name.is_empty() {
        return Err("name field is required and must not be empty".into());
    }

    // Validate allowed_tools list entries are non-empty.
    if let AllowedTools::List(tools) = &fm.allowed_tools {
        if tools.is_empty() {
            warnings.push("allowed_tools list is empty — skill will have no tools".to_string());
        }
        for t in tools {
            if t.is_empty() {
                warnings.push("allowed_tools contains an empty string".to_string());
            }
        }
    }

    // Validate hook declarations.
    for (event, hooks) in &fm.hooks {
        for hook in hooks {
            if hook.name.is_empty() {
                warnings.push(format!("hook in event '{event}' has empty name"));
            }
            if hook.run.is_empty() {
                warnings.push(format!(
                    "hook '{}' in event '{event}' has empty run command",
                    hook.name
                ));
            }
            // handler_type is validated by serde deserialization.
            let _ = hook.handler_type; // suppress unused warning
        }
    }

    if warnings.is_empty() {
        Ok(())
    } else {
        Err(warnings.join("; "))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::HookHandlerType;

    #[test]
    fn parse_full_frontmatter() {
        let content = r#"---
name: commit
description: Create a git commit
type: prompt
allowed_tools:
  - Bash
  - FileRead
  - Grep
model: sonnet
hooks:
  PostToolUse:
    - name: verify-commit
      type: bash
      if: "tool == 'Bash'"
      run: "echo ok"
trigger: null
---
Do the commit with $ARGS
"#;
        let result = parse_skill_file(content, Path::new("commit.md"));
        let skill = result.skill().expect("should parse");
        assert_eq!(skill.frontmatter.name, "commit");
        assert_eq!(skill.frontmatter.skill_type, SkillType::Prompt);
        assert_eq!(
            skill.frontmatter.allowed_tools,
            AllowedTools::list(vec!["Bash".into(), "FileRead".into(), "Grep".into()])
        );
        assert_eq!(skill.frontmatter.model.as_deref(), Some("sonnet"));
        assert!(skill.frontmatter.hooks.contains_key("PostToolUse"));
        assert_eq!(skill.body.trim(), "Do the commit with $ARGS");
    }

    #[test]
    fn parse_backward_compat_no_frontmatter() {
        let content = "Just a plain markdown prompt\nwith $ARGS placeholder";
        let result = parse_skill_file(content, Path::new("/dir/review.md"));
        let skill = result.skill().expect("should parse");
        assert_eq!(skill.frontmatter.name, "review");
        assert_eq!(skill.frontmatter.skill_type, SkillType::Prompt);
        assert!(skill.frontmatter.allowed_tools.is_all());
    }

    #[test]
    fn parse_invalid_yaml_returns_error() {
        let content = "---\n[invalid yaml\n---\nbody";
        let result = parse_skill_file(content, Path::new("bad.md"));
        assert!(result.is_error());
    }

    #[test]
    fn parse_missing_name_returns_error() {
        let content = "---\ndescription: no name field\n---\nbody";
        let result = parse_skill_file(content, Path::new("noname.md"));
        // serde_yaml requires 'name' field → this is a deserialization error
        assert!(result.is_error());
    }

    #[test]
    fn parse_agent_type() {
        let content = "---\nname: deploy\ntype: agent\n---\nDeploy $ARGS";
        let result = parse_skill_file(content, Path::new("deploy.md"));
        let skill = result.skill().expect("should parse");
        assert_eq!(skill.frontmatter.skill_type, SkillType::Agent);
    }

    #[test]
    fn parse_headless_type() {
        let content = "---\nname: lint\ntype: headless\n---\nLint the code";
        let result = parse_skill_file(content, Path::new("lint.md"));
        let skill = result.skill().expect("should parse");
        assert_eq!(skill.frontmatter.skill_type, SkillType::Headless);
    }

    #[test]
    fn render_prompt_replaces_args() {
        let content = "---\nname: test\n---\nRun tests for $ARGS please";
        let result = parse_skill_file(content, Path::new("test.md"));
        let skill = result.skill().unwrap();
        assert_eq!(
            skill.render_prompt("auth module"),
            "Run tests for auth module please"
        );
    }

    #[test]
    fn skill_name_from_path_strips_extension() {
        assert_eq!(
            skill_name_from_path(Path::new("/foo/bar/my-skill.md")),
            "my-skill"
        );
    }

    #[test]
    fn empty_allowed_tools_list_produces_warning() {
        let content = "---\nname: empty\nallowed_tools: []\n---\nbody";
        let result = parse_skill_file(content, Path::new("empty.md"));
        assert!(matches!(result, SkillLoadResult::Warning { .. }));
    }

    #[test]
    fn hook_with_all_handler_types() {
        let content = r#"---
name: multi
hooks:
  PreToolUse:
    - name: bash-hook
      type: bash
      run: "echo pre"
  PostToolUse:
    - name: agent-hook
      type: agent
      run: "check output"
  UserPromptSubmit:
    - name: prompt-hook
      type: prompt
      run: "validate prompt"
---
body
"#;
        let result = parse_skill_file(content, Path::new("multi.md"));
        let skill = result.skill().expect("should parse");
        assert_eq!(skill.frontmatter.hooks.len(), 3);
        let pre = &skill.frontmatter.hooks["PreToolUse"][0];
        assert_eq!(pre.handler_type, HookHandlerType::Bash);
        let post = &skill.frontmatter.hooks["PostToolUse"][0];
        assert_eq!(post.handler_type, HookHandlerType::Agent);
    }
}
